cdc/sink/mq.go

```diff
@@ -322,37 +319,62 @@ func (k *mqSink) runWorker(ctx context.Context, partition int32) error {
 		case <-ctx.Done():
 			return ctx.Err()
 		case <-tick.C:
-			if err := flushToProducer(); err != nil {
+			if err := flushToProducer(codec.EncoderNeedAsyncWrite); err != nil {
 				return errors.Trace(err)
 			}
 			continue
 		case e = <-input:
 		}
 		if e.row == nil {
 			if e.resolvedTs != 0 {
-				if err := flushToProducer(); err != nil {
+				if err := flushToProducer(codec.EncoderNeedAsyncWrite); err != nil {
 					return errors.Trace(err)
 				}
 				atomic.StoreUint64(&k.partitionResolvedTs[partition], e.resolvedTs)
 				k.resolvedNotifier.Notify()
 			}
 			continue
 		}
-		err := encoder.AppendRowChangedEvent(e.row)
+		op, err := encoder.AppendRowChangedEvent(e.row)
 		if err != nil {
 			return errors.Trace(err)
 		}
 		batchSize++
 		// This is only a temporary fix so that the Avro encoder does not overflow.
 		// Pending further refactoring
-		if encoder.Size() >= batchSizeLimit || (k.protocol == codec.ProtocolAvro && encoder.Size() >= 1) {
-			if err := flushToProducer(); err != nil {
+		if encoder.Size() >= batchSizeLimit || op == codec.EncoderNeedSyncWrite || op == codec.EncoderNeedAsyncWrite {
```

Comment:
Currently, op is 
- `codec.EncoderNeedAsyncWrite` if encoder is `AvroEncoder`
- `codec.EncoderNoOperation` if encoder is other encoder

So when `encoder.Size() >= batchSizeLimit` is true and encoder is not `AvroEncoder`, the `flushToProducer` will output a warning log and does nothing. The real `send message` will have to wait for flush ticker.